### ISSUE TITLE
Player fails to play some flac files

### ISSUE THREAD

**Issue by Porkepix** [OxygenCobalt/Auxio#619]
### Describe the Bug/Crash

When arriving on some flac files through random play, it just skip it and go directly to the next (appears on screen a split second).

### Describe the intended behavior

It should play it normally like other players can.

### What android version do you use?

Android 8 (Oreo)

### What device model do you use?

Sony Xperia XZ

### Bug report

Relevant `adb logcat -v threadtime` logs:

```
11-20 11:19:42.583  1430  2159 I MediaFocusControl:  AudioFocus  requestAudioFocus() from uid/pid 10335/31485 clientId=android.media.AudioManager@e16e6e8androidx.media3.exoplayer.AudioFocusManager$AudioFocusListener@98b3101 req=1 flags=0x0
11-20 11:19:42.594  1050  1318 D NetlinkEvent: Unknown ND option type 31
11-20 11:19:42.790  2013  2013 W StaticLayout: maxLineHeight should not be -1.  maxLines:1 lineCount:1
11-20 11:19:42.791  2013  2013 I chatty  : uid=10076(u0_a76) com.android.systemui identical 2 lines
11-20 11:19:42.791  2013  2013 W StaticLayout: maxLineHeight should not be -1.  maxLines:1 lineCount:1
11-20 11:19:42.859   665   918 W qti_sensors_hal: Thresh::processInd: Proximity Status 0.0, A/D 0xb2223550(2988586320)
11-20 11:19:42.886 31485  7719 I OMXClient: Treble IOmx obtained
11-20 11:19:42.888  1053  1249 I OMXMaster: makeComponentInstance(OMX.qti.audio.decoder.flac) in omx@1.0-service process
11-20 11:19:42.967 31485 31496 I zygote64: Background concurrent copying GC freed 197684(5MB) AllocSpace objects, 13(400KB) LOS objects, 45% free, 28MB/52MB, paused 4.208ms total 155.232ms
11-20 11:19:43.018  1053  1249 E OMXNodeInstance: setConfig(0xe9823980:qti.decoder.flac, ConfigPriority(0x6f800002)) ERROR: Undefined(0x80001001)
11-20 11:19:43.018 31485  7719 I ACodec  : codec does not support config priority (err -2147483648)
11-20 11:19:43.022 31485  7719 I ExtendedACodec: no bitwidth, setting default bitwidth as 16 bits
11-20 11:19:43.022 31485  7719 I ExtendedACodec: no min blksize, setting default block size as 16
11-20 11:19:43.022 31485  7719 I ExtendedACodec: no max blksize, setting default block size as 16
11-20 11:19:43.022 31485  7719 I ExtendedACodec: no min frame size, setting default frame size as 0
11-20 11:19:43.022 31485  7719 I ExtendedACodec: no max frame size, setting default frame size as 0
11-20 11:19:43.023  1053 30126 D SoftFlacDecoder: CSD: nChannels=2, nSampleRate=48000, nBitsPerSample=16
11-20 11:19:43.023  1053 30126 D SoftFlacDecoder: CSD: nMinBlkSize=16, nMaxBlkSize=16
11-20 11:19:43.023  1053 30126 D SoftFlacDecoder: CSD: nMinFrmSize=0, nMaxFrmSize=0
11-20 11:19:43.030  1053  1249 E OMXNodeInstance: setConfig(0xe9823980:qti.decoder.flac, ConfigPriority(0x6f800002)) ERROR: Undefined(0x80001001)
11-20 11:19:43.030 31485  7719 I ACodec  : codec does not support config priority (err -2147483648)
11-20 11:19:43.030  1053 30126 E OMXNodeInstance: setConfig(0xe9823980:qti.decoder.flac, ConfigOperatingRate(0x6f800003)) ERROR: Undefined(0x80001001)
11-20 11:19:43.030 31485  7719 I ACodec  : codec does not support config operating rate (err -2147483648)
11-20 11:19:43.033  1053  1249 E OMXNodeInstance: getConfig(0xe9823980:qti.decoder.flac, ConfigAndroidVendorExtension(0x6f100004)) ERROR: Undefined(0x80001001)
11-20 11:19:43.034 31485  7719 W ExtendedACodec: Failed to get extension for extradata parameter
11-20 11:19:43.148  1053  7720 E SoftFlacDecoder: inSize (20816) > avail size (17427)
11-20 11:19:43.149 31485  7719 E ACodec  : [OMX.qti.audio.decoder.flac] ERROR(0x80001001)
11-20 11:19:43.149 31485  7719 E ACodec  : signalError(omxError 0x80001001, internalError -2147483648)
11-20 11:19:43.149 31485  7719 E MediaCodec: Codec reported err 0x80001001, actionCode 0, while in state 6
11-20 11:19:43.157 31485 31512 E MediaCodecAudioRenderer: Audio codec error
11-20 11:19:43.157 31485 31512 E MediaCodecAudioRenderer:   java.lang.IllegalStateException
11-20 11:19:43.157 31485 31512 E MediaCodecAudioRenderer:       at android.media.MediaCodec.native_dequeueOutputBuffer(Native Method)
11-20 11:19:43.157 31485 31512 E MediaCodecAudioRenderer:       at android.media.MediaCodec.dequeueOutputBuffer(MediaCodec.java:2637)
11-20 11:19:43.157 31485 31512 E MediaCodecAudioRenderer:       at androidx.media3.exoplayer.mediacodec.SynchronousMediaCodecAdapter.dequeueOutputBufferIndex(Unknown Source:4)
11-20 11:19:43.157 31485 31512 E MediaCodecAudioRenderer:       at androidx.media3.exoplayer.audio.MediaCodecAudioRenderer.drainOutputBuffer(SourceFile:46)
11-20 11:19:43.157 31485 31512 E MediaCodecAudioRenderer:       at androidx.media3.exoplayer.audio.MediaCodecAudioRenderer.render(SourceFile:73)
11-20 11:19:43.157 31485 31512 E MediaCodecAudioRenderer:       at androidx.media3.exoplayer.ExoPlayerImplInternal.doSomeWork(SourceFile:89)
11-20 11:19:43.157 31485 31512 E MediaCodecAudioRenderer:       at androidx.media3.exoplayer.ExoPlayerImplInternal.handleMessage(SourceFile:289)
11-20 11:19:43.157 31485 31512 E MediaCodecAudioRenderer:       at android.os.Handler.dispatchMessage(Handler.java:101)
11-20 11:19:43.157 31485 31512 E MediaCodecAudioRenderer:       at android.os.Looper.loop(Looper.java:251)
11-20 11:19:43.157 31485 31512 E MediaCodecAudioRenderer:       at android.os.HandlerThread.run(HandlerThread.java:65)
11-20 11:19:43.159 31485 31512 E ExoPlayerImplInternal: Playback error
11-20 11:19:43.159 31485 31512 E ExoPlayerImplInternal:   androidx.media3.exoplayer.ExoPlaybackException: MediaCodecAudioRenderer error, index=0, format=Format(null, null, null, audio/flac, null, -1, null, [-1, -1, -1.0, null], [2, 48000]), format_supported=YES
11-20 11:19:43.159 31485 31512 E ExoPlayerImplInternal:       at androidx.media3.exoplayer.ExoPlayerImplInternal.handleMessage(SourceFile:403)
11-20 11:19:43.159 31485 31512 E ExoPlayerImplInternal:       at android.os.Handler.dispatchMessage(Handler.java:101)
11-20 11:19:43.159 31485 31512 E ExoPlayerImplInternal:       at android.os.Looper.loop(Looper.java:251)
11-20 11:19:43.159 31485 31512 E ExoPlayerImplInternal:       at android.os.HandlerThread.run(HandlerThread.java:65)
11-20 11:19:43.159 31485 31512 E ExoPlayerImplInternal:   Caused by: androidx.media3.exoplayer.mediacodec.MediaCodecDecoderException: Decoder failed: OMX.qti.audio.decoder.flac
11-20 11:19:43.159 31485 31512 E ExoPlayerImplInternal:       at androidx.media3.exoplayer.audio.MediaCodecAudioRenderer.render(SourceFile:242)
11-20 11:19:43.159 31485 31512 E ExoPlayerImplInternal:       at androidx.media3.exoplayer.ExoPlayerImplInternal.doSomeWork(SourceFile:89)
11-20 11:19:43.159 31485 31512 E ExoPlayerImplInternal:       at androidx.media3.exoplayer.ExoPlayerImplInternal.handleMessage(SourceFile:289)
11-20 11:19:43.159 31485 31512 E ExoPlayerImplInternal:       ... 3 more
11-20 11:19:43.159 31485 31512 E ExoPlayerImplInternal:   Caused by: java.lang.IllegalStateException
11-20 11:19:43.159 31485 31512 E ExoPlayerImplInternal:       at android.media.MediaCodec.native_dequeueOutputBuffer(Native Method)
11-20 11:19:43.159 31485 31512 E ExoPlayerImplInternal:       at android.media.MediaCodec.dequeueOutputBuffer(MediaCodec.java:2637)
11-20 11:19:43.159 31485 31512 E ExoPlayerImplInternal:       at androidx.media3.exoplayer.mediacodec.SynchronousMediaCodecAdapter.dequeueOutputBufferIndex(Unknown Source:4)
11-20 11:19:43.159 31485 31512 E ExoPlayerImplInternal:       at androidx.media3.exoplayer.audio.MediaCodecAudioRenderer.drainOutputBuffer(SourceFile:46)
11-20 11:19:43.159 31485 31512 E ExoPlayerImplInternal:       at androidx.media3.exoplayer.audio.MediaCodecAudioRenderer.render(SourceFile:73)
11-20 11:19:43.159 31485 31512 E ExoPlayerImplInternal:       ... 5 more
11-20 11:19:43.161 31485 31512 E ExoPlayerImplInternal: Disable failed.
11-20 11:19:43.161 31485 31512 E ExoPlayerImplInternal:   java.lang.IllegalStateException
11-20 11:19:43.161 31485 31512 E ExoPlayerImplInternal:       at android.media.MediaCodec.native_flush(Native Method)
11-20 11:19:43.161 31485 31512 E ExoPlayerImplInternal:       at android.media.MediaCodec.flush(MediaCodec.java:2121)
11-20 11:19:43.161 31485 31512 E ExoPlayerImplInternal:       at androidx.media3.exoplayer.mediacodec.SynchronousMediaCodecAdapter.flush(Unknown Source:2)
11-20 11:19:43.161 31485 31512 E ExoPlayerImplInternal:       at androidx.media3.exoplayer.audio.MediaCodecAudioRenderer.flushCodec(SourceFile:3)
11-20 11:19:43.161 31485 31512 E ExoPlayerImplInternal:       at androidx.media3.exoplayer.audio.MediaCodecAudioRenderer.flushOrReleaseCodec(SourceFile:64)
11-20 11:19:43.161 31485 31512 E ExoPlayerImplInternal:       at androidx.media3.exoplayer.audio.MediaCodecAudioRenderer.onDisabled(SourceFile:28)
11-20 11:19:43.161 31485 31512 E ExoPlayerImplInternal:       at androidx.media3.exoplayer.ExoPlayerImplInternal.disableRenderer(SourceFile:65)
11-20 11:19:43.161 31485 31512 E ExoPlayerImplInternal:       at androidx.media3.exoplayer.ExoPlayerImplInternal.resetInternal(SourceFile:53)
11-20 11:19:43.161 31485 31512 E ExoPlayerImplInternal:       at androidx.media3.exoplayer.ExoPlayerImplInternal.stopInternal(SourceFile:13)
11-20 11:19:43.161 31485 31512 E ExoPlayerImplInternal:       at androidx.media3.exoplayer.ExoPlayerImplInternal.handleMessage(SourceFile:520)
11-20 11:19:43.161 31485 31512 E ExoPlayerImplInternal:       at android.os.Handler.dispatchMessage(Handler.java:101)
11-20 11:19:43.161 31485 31512 E ExoPlayerImplInternal:       at android.os.Looper.loop(Looper.java:251)
11-20 11:19:43.161 31485 31512 E ExoPlayerImplInternal:       at android.os.HandlerThread.run(HandlerThread.java:65)
```

### Duplicates

- [X] I have checked the [Troubleshooting](https://github.com/OxygenCobalt/Auxio/wiki/Troubleshooting) page.
- [X] I have checked this issue for duplicates.
- [X] I have checked that this issue occurs on the [lastest version](https://github.com/OxygenCobalt/Auxio/releases).
- [X] I agree to the [Contribution Guidelines](https://github.com/OxygenCobalt/Auxio/blob/dev/.github/CONTRIBUTING.md).

**Comment by OxygenCobalt** [comment]
Android 8 doesn't support FLAC normally, but the FFMpeg extension should be the one decoding them in Auxio then. Confused at why this is occurring, let me investigate.

**Comment by OxygenCobalt** [comment]
Okay, that may imply an error on ExoPlayer's end. Can I have a sample file at @Porkepix?

**Comment by OxygenCobalt** [comment]
That's a consequence of the app just giving up and skipping to the next track if it fails to play something. It's not a bug as much a consequence of a stupid hack. I want to improve it, but I haven't had any time.

**Comment by OxygenCobalt** [comment]
Also, just because the logs are flooded doesn't mean it must be a bug. The logs are always flooded with information when you use any app.

**Comment by OxygenCobalt** [comment]
> Also, just because the logs are flooded doesn't mean it must be a bug. The logs are always flooded with information when you use any app.

You underestimate how complex (and arguably bloated) the android system is.

> What is the hack here, the skip, of the fact it retries playing the whole album less than a minute later?

The skip. It restarts playing the whole album because Auxio prepares audio files for rendering regardless of if playback is going on. This causes it to error out from this FLAC issue, and Auxio begins skipping again. Again, it's incredibly stupid, but a better system requires setting up notifications, improving internal state management, etc. I'm lucky if I'm able to modify more than a few lines of code every couple of months with the obligations I currently have.

**Comment by OxygenCobalt** [comment]
Alright, make the issue. I think I was mis-remembering making an issue in the past when I really had a reminder somewhere.

**Comment by OxygenCobalt** [comment]
> Should it not skip in case of an error such as the FLAC one here? (result in more smooth usage, but also make issue much less visible: it was out of pure randomness I noticed the FLAC issue because I was looking at the screen at the same timing); if it should not skip, what should be the behavior?

The behavior should be that playback stops, and then the app indicates openly that an error occurred. This is the problem, there would need to be a notification if background playback was occurring, and also a UI indicator in-app that says "Failed to play this file", alongside a "More" button that shows a stack trace so I can debug it.

> When it finished to try to play the whole album and failed on the last track, should it retry playing the same album a couple of dozens seconds later? Imho I don't think it should: playback stopped, I see no reason it should restart plays by itself, but from my reading of your comment, you seemed to think otherwise.

No, it should not. Removing the skipping should also fix this.

**Comment by OxygenCobalt** [comment]
Update: I can't reproduce this on a virtual device, sadly. I think it's specific to your Xperia device. I'm going to have to forward you to the ExoPlayer maintainers themselves. Be sure to provide the file causing the issue, the stack trace, and a link to this issue @Porkepix. Really sorry about this.

**Comment by OxygenCobalt** [comment]
> Your link leads to https://github.com/androidx/media while https://github.com/google/ExoPlayer/ exists;

ExoPlayer moved to media3. I use the terms interchangably.

> That being said, a little research lead me to https://github.com/google/ExoPlayer/issues/8261 and it looks a little bit similar; some comments seems to suggest the use of an extension…?

Huh. I already use an alternative decoder, but it's not being used for some reason. Can you try this apk @Porkepix?

[Auxio_PrioritizeFFMpeg.zip](https://github.com/OxygenCobalt/Auxio/files/13694184/Auxio_PrioritizeFFMpeg.zip)

**Comment by OxygenCobalt** [comment]
> So, appart from some things looking a bit different (but I guess it's normal/intended), such as notification icon being different during indexing and some notification about heap dump that's now gone and I didn't screenshot (it also looks to have installed another "leak" application), I tried several songs that weren't playing and all of them now plays (I didn't tried every of them as I didn't referenced every problematic ones), so I'm guessing the thing you tried here fixes the issue.

It's a debug build that has some extra development stuff (Such as the kinda unreliable memory leak detector) and some unreleased features I'm still trying to pull together. Great that it works though.

> So now for things to be clear to me: did you change to another decoder that's not the one you're usually using, or is it still using the same one, but you found some fixes about the player not picking up the right one in every case?
In case you changed the decoder, do you plan on using that one instead, or was it just to confirm some decoder issue? And like, what are the possible decoders and their respective pro and cons?

Basically, I just made it so that Auxio always uses ffmpeg to decode FLAC files. I was under the impression ExoPlayer would fall back to ffmpeg prior, but this error occurs too late when loading to allow that to happen. ffmpeg is really reliable, so this should rule out any chance of these issues in the future, at least with flac files.

@Porkepix

**Comment by OxygenCobalt** [comment]
> So, ExoPlayer in one possible decoder, and ffmpeg is another one? Is there some others than those two? And what are their respective pro/cons?

Theres MediaCodec, which is whatever Android provides and is unreliable apparently. Ffmpeg is another, it's more reliable and feature-rich, yet also somewhat large. There is also libFLAC and libOPUS, which are useful for reducing size if you only want support for those particular files. For my purposes, I need to support AAC and FLAC, and only ffmpeg allows me to do that.

> And I guess ExoPlayer decoder is still having an issue then, but that this should be reported upstream? If so, which repo is best fit for that? Unless the issue I linked is the exact same one and in that case there's no need to create a dupe.

Yeah, this issue is intractable on ExoPlayer's end. It's just a problem with your phone's default FLAC encoder. Other apps don't have this issue since MediaPlayer has lower-level system access and can thus circumvent these issues.

**Comment by OxygenCobalt** [comment]
> What's unreliable about MediaCodec?

It relies on whatever your OEM (ex. Sony) provides, which might have strange issues. In this case, your decoder, QTI has been known to produce these issues.

> Also, your explanation gave me an idea; the phone still have the default preinstalled Sony's music player (you can't uninstall it).
Code is of course proprietary, but I would guess it uses what you're thinking about when you talk about the default FLAC decoder the phone provides?
So I checked how was this player reacting to those albums. Interestingly, the player plays those albums without any issues, but the exact same albums Auxio had trouble with also have some "HR" logo on top of their cover which I would bet stands for "High Res Audio" (it's not explained anywhere).

I'm honestly not sure, but Sony might be self-rolling their own audio player (including their own decoder) if it doesn't have the issue (it is Sony...). Some other players do that.
